pkg/storegateway/indexheader/lazy_binary_reader.go

```diff
@@ -230,125 +268,170 @@ func (r *LazyBinaryReader) EagerLoad() {
 	// then goroutine that continues would not be able to get Write lock.)
 	err := r.lazyLoadingGate.Start(r.ctx)
 	if err != nil {
-		return errors.Wrapf(err, "failed to wait for turn")
+		return nil, errors.Wrapf(err, "failed to wait for turn")
 	}
 	defer r.lazyLoadingGate.Done()
 
-	r.readerMx.Lock()
-	defer r.readerMx.Unlock()
-
-	// Ensure none else tried to load it in the meanwhile.
-	if r.reader != nil {
-		return nil
-	}
-	if r.readerErr != nil {
-		return r.readerErr
-	}
-
 	level.Debug(r.logger).Log("msg", "lazy loading index-header file", "path", r.filepath)
 	r.metrics.loadCount.Inc()
 	startTime := time.Now()
 
 	reader, err := r.readerFactory()
 	if err != nil {
 		r.metrics.loadFailedCount.Inc()
-		r.readerErr = err
-		return errors.Wrapf(err, "lazy load index-header file at %s", r.filepath)
+		return nil, errors.Wrapf(err, "lazy load index-header file at %s", r.filepath)
 	}
 
-	r.reader = reader
 	elapsed := time.Since(startTime)
 
 	level.Debug(r.logger).Log("msg", "lazy loaded index-header file", "path", r.filepath, "elapsed", elapsed)
 	r.metrics.loadDuration.Observe(elapsed.Seconds())
 
-	return nil
+	return reader, nil
 }
 
-// unloadIfIdleSince closes underlying BinaryReader if the reader is idle since given time (as unix nano). If idleSince is 0,
-// the check on the last usage is skipped. Calling this function on a already unloaded reader is a no-op.
-func (r *LazyBinaryReader) unloadIfIdleSince(ts int64) error {
-	r.readerMx.Lock()
-	defer r.readerMx.Unlock()
-
-	// Nothing to do if already unloaded.
-	if r.reader == nil {
-		return nil
-	}
-
-	// Do not unloadIfIdleSince if not idle.
-	if ts > 0 && r.usedAt.Load() > ts {
-		return errNotIdle
+func (r *LazyBinaryReader) waitAndCloseReader(req readerRequest) {
+	resp := <-req.response
+	if resp.reader != nil {
+		resp.inUse.Done()
 	}
+}
 
-	// Wait until all users finished using current reader.
-	r.readerInUse.Wait()
-
-	r.metrics.unloadCount.Inc()
-	if err := r.reader.Close(); err != nil {
-		r.metrics.unloadFailedCount.Inc()
-		return err
+// unloadIfIdleSince closes underlying BinaryReader if the reader is idle since given time (as unix nano). If idleSince is 0,
+// the check on the last usage is skipped. Calling this function on a already unloaded reader is a no-op.
+func (r *LazyBinaryReader) unloadIfIdleSince(tsNano int64) error {
+	req := unloadRequest{
+		// The channel is unbuffered because we will read the response. It should be buffered if we can give up before reading from it
+		response:       make(chan error),
+		idleSinceNanos: tsNano,
 	}
-
-	r.reader = nil
-	return nil
+	r.unloadReqC <- req
+	return <-req.response
 }
 
-// isIdleSince returns true if the reader is idle since given time (as unix nano).
-func (r *LazyBinaryReader) isIdleSince(ts int64) bool {
-	if r.usedAt.Load() > ts {
-		return false
+func (r *LazyBinaryReader) controlLoop() {
+	var loaded loadedReader
+	for {
+		select {
+		case readerReq := <-r.loadedReaderC:
```

Comment:
the only serial part here is obtaining a reader. You can obtain multiple readers and use them parallel.

Granted, it will be slower to obtain 100 readers because you need to do 200 channel send+receive serially compared to the 100 concurrent RLock+RUnlocks before. But my gut feeling is that it is negligible in the scheme of the i/o-intensive work that the store-gateway does.

Reply:
i don't think we'll be A-B testing this, so I can try to run a few benchmarks before merging

Reply:
> Granted, it will be slower to obtain 100 readers because you need to do 200 channel send+receive serially compared to the 100 concurrent RLock+RUnlocks before.

I think we're talking nanoseconds (maybe a microsecond...) here., it shouldn't be an issue.

Reply:
realistically it takes between 0.5 and 23 microseconds to acquire a reader. A single-threaded access is the slowest - 67ns -> 490ns. Generally, the higher the concurrency, the lower the difference between the two approaches

```
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/storegateway/indexheader
                                                  │    before    │                after                 │
                                                  │    sec/op    │    sec/op     vs base                │
NewLazyBinaryReaderConcurrency/concurrency=1-10     67.23n ± 15%   490.40n ± 2%  +629.38% (p=0.002 n=6)
NewLazyBinaryReaderConcurrency/concurrency=2-10     238.7n ±  2%    903.4n ± 2%  +278.55% (p=0.002 n=6)
NewLazyBinaryReaderConcurrency/concurrency=10-10    1.490µ ±  2%    4.527µ ± 1%  +203.83% (p=0.002 n=6)
NewLazyBinaryReaderConcurrency/concurrency=20-10    2.960µ ± 22%    9.269µ ± 0%  +213.18% (p=0.002 n=6)
NewLazyBinaryReaderConcurrency/concurrency=50-10    8.880µ ± 19%   23.920µ ± 1%  +169.38% (p=0.002 n=6)
NewLazyBinaryReaderConcurrency/concurrency=100-10   18.00µ ±  4%    47.51µ ± 1%  +164.01% (p=0.002 n=6)
geomean                                             1.498µ          5.258µ       +250.95%
```

I also ran a modified version of `BenchmarkLabelNames` - one which uses a lazy reader instead of a plain reader. Because that method is so fast we end up with a significant slowdown when using the changes on this branch


```
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/storegateway/indexheader
                                 │ before-labelnames-with-lazy-reader │   after-labelnames-with-lazy-reader   │
                                 │               sec/op               │    sec/op      vs base                │
LabelNames/20Names100Values-10                           1.554µ ±  2%    2.763µ ±  1%   +77.82% (p=0.002 n=6)
LabelNames/20Names500Values-10                           1.561µ ± 17%    3.334µ ±  4%  +113.58% (p=0.002 n=6)
LabelNames/20Names1000Values-10                          1.551µ ±  1%    3.388µ ±  2%  +118.44% (p=0.002 n=6)
LabelNames/50Names100Values-10                           3.954µ ± 14%    7.139µ ±  2%   +80.55% (p=0.002 n=6)
LabelNames/50Names500Values-10                           3.927µ ±  4%    7.054µ ± 12%   +79.64% (p=0.002 n=6)
LabelNames/50Names1000Values-10                          3.875µ ±  3%    5.289µ ± 16%   +36.49% (p=0.002 n=6)
LabelNames/100Names100Values-10                          8.496µ ±  1%   10.402µ ±  1%   +22.43% (p=0.002 n=6)
LabelNames/100Names500Values-10                          8.493µ ±  1%   10.171µ ±  2%   +19.74% (p=0.002 n=6)
LabelNames/100Names1000Values-10                         8.507µ ±  1%   10.015µ ± 12%   +17.73% (p=0.002 n=6)
LabelNames/200Names100Values-10                          18.11µ ±  0%    20.57µ ±  5%   +13.56% (p=0.002 n=6)
LabelNames/200Names500Values-10                          18.09µ ±  1%    20.50µ ±  1%   +13.29% (p=0.002 n=6)
LabelNames/200Names1000Values-10                         18.15µ ±  1%    20.44µ ±  1%   +12.61% (p=0.002 n=6)
geomean                                                  5.535µ          8.066µ         +45.74%
```

Reply:
the stats are similar for `BenchmarkLabelValuesOffsetsIndexV2` with a lazy reader - 9% slowdown.

Reply:
i couldn't come up with something that's not terribly overcomplicated. Happy to jump on a call with one of your if you're curious what I tried.

I think we should be able to stomach this slow-down. It should only show up for already fast methods. This is the higher-level `BenchmarkBucketStoreLabelValues` which tries to simulate a full store-gateway. The results are mixes

<details><summary>Details</summary>
<p>

```
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/storegateway
                                                                                                                                         │ before-highlevel-benchmarks │      after-highlevel-benchmarks      │
                                                                                                                                         │           sec/op            │    sec/op     vs base                │
BucketStoreLabelValues/no_cache/10-series-matched-with-10-label-values-10                                                                                 658.0µ ±  1%   753.4µ ±  4%  +14.50% (p=0.000 n=12)
BucketStoreLabelValues/no_cache/1000-series-matched-with-1000-label-values-10                                                                             16.27m ±  3%   20.92m ± 14%  +28.57% (p=0.000 n=12)
BucketStoreLabelValues/no_cache/1_000_000-series-matched-with-10-label-values-10                                                                          29.86m ±  2%   29.66m ±  6%        ~ (p=0.799 n=12)
BucketStoreLabelValues/no_cache/1_000_000-series-matched-with-1-label-values-10                                                                           217.9µ ±  7%   239.6µ ±  1%   +9.98% (p=0.000 n=12)
BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/10-series-matched-with-10-label-values-10                                              118.5µ ±  4%   124.3µ ±  6%   +4.87% (p=0.007 n=12)
BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1000-series-matched-with-1000-label-values-10                                          1.127m ±  2%   1.181m ± 12%   +4.79% (p=0.002 n=12)
BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1_000_000-series-matched-with-10-label-values-10                                       29.52m ±  1%   30.42m ±  9%   +3.07% (p=0.000 n=12)
BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1_000_000-series-matched-with-1-label-values-10                                        94.45µ ±  1%   99.96µ ±  2%   +5.83% (p=0.000 n=12)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/1of1000000-10                                     129.6m ±  7%   126.4m ±  2%   -2.46% (p=0.015 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/10of1000000-10                                    127.0m ±  1%   125.7m ±  1%   -1.00% (p=0.009 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/1000000of1000000-10                                1.473 ±  3%    1.493 ±  6%        ~ (p=0.240 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/1of1000000-10                           209.2m ± 27%   202.7m ±  4%        ~ (p=0.065 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/10of1000000-10                          234.3m ± 41%   199.5m ±  4%  -14.84% (p=0.004 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/1000000of1000000-10                      1.841 ± 47%    1.520 ± 25%  -17.44% (p=0.041 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/1of1000000-10                                      151.5m ±  9%   127.9m ±  1%  -15.61% (p=0.002 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/10of1000000-10                                     146.5m ± 14%   127.0m ±  2%  -13.33% (p=0.002 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/1000000of1000000-10                                 1.530 ± 37%    1.512 ±  5%        ~ (p=0.589 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/1of10000000-10                                   14.03m ±  3%   13.82m ±  1%   -1.50% (p=0.015 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/100of10000000-10                                 14.05m ± 22%   13.78m ±  2%   -1.95% (p=0.009 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/10000000of10000000-10                            176.7m ± 17%   174.2m ± 11%        ~ (p=0.699 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/1of10000000-10                         19.50m ±  1%   19.08m ±  6%        ~ (p=0.093 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/100of10000000-10                       19.19m ±  2%   19.31m ± 17%        ~ (p=0.699 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/10000000of10000000-10                  170.4m ±  2%   173.0m ±  3%        ~ (p=0.093 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/1of10000000-10                                    13.74m ±  1%   14.33m ±  3%   +4.26% (p=0.002 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/100of10000000-10                                  13.68m ±  4%   14.92m ± 15%   +9.05% (p=0.002 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/10000000of10000000-10                             173.7m ± 27%   168.7m ±  6%        ~ (p=1.000 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/1of10000000-10                                   2.174m ± 10%   2.321m ±  4%   +6.75% (p=0.015 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/100of10000000-10                                 2.279m ± 14%   2.379m ± 25%        ~ (p=0.240 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/10000000of10000000-10                            89.41m ±  3%   88.06m ± 15%        ~ (p=0.485 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/1of10000000-10                         2.267m ± 11%   2.194m ±  4%        ~ (p=0.240 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/100of10000000-10                       2.224m ±  6%   2.216m ±  3%        ~ (p=0.937 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/10000000of10000000-10                  86.48m ±  3%   86.55m ±  2%        ~ (p=0.937 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/1of10000000-10                                    2.499m ±  6%   2.332m ± 34%        ~ (p=0.394 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/100of10000000-10                                  2.473m ±  6%   2.353m ±  7%        ~ (p=0.240 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/10000000of10000000-10                             88.41m ±  7%   87.52m ±  2%        ~ (p=0.485 n=6)
geomean                                                                                                                                                   20.85m         20.79m         -0.27%

```

</p>
</details>

Reply:
I think we can ignore this difference.